src/cobra/core/gene.py

```diff
@@ -122,6 +128,96 @@ def parse_gpr(str_expr: str) -> Tuple:
     return gpr_tree, gpr_tree.genes
 
 
+class GPRSympifier(NodeVisitor):
+    """Parses compiled ast of a gene_reaction_rule to sympy and identifies genes"""
+
+    def __init__(self):
+        NodeVisitor.__init__(self)
+
+    def visit_Expression(self, node):
+        return self.visit(node.body) if hasattr(node, "body") else None
+
+    def visit_GPR(self, node):
+        return self.visit(node.body) if hasattr(node, "body") else None
+
+    def visit_Name(self, node):
+        if node.id.startswith("__cobra_escape__"):
+            node.id = node.id[16:]
+        for char, escaped in replacements:
+            if escaped in node.id:
+                node.id = node.id.replace(escaped, char)
+        return Symbol(node.id)
+
+    def visit_BinOp(self, node):
+        if isinstance(node.op, BitAnd):
+            return sp_And(*[node.left, node.right])
+        elif isinstance(node.op, BitOr):
+            return sp_Or(*[node.left, node.right])
+        else:
+            raise TypeError("unsupported operation '%s'" % node.op.__class__.__name__)
+
+    def visit_BoolOp(self, node):
+        if isinstance(node.op, And):
+            return sp_And(*[self.visit(i) for i in node.values])
+        if isinstance(node.op, Or):
+            return sp_Or(*[self.visit(i) for i in node.values])
+
+
+def parse_gpr_sympy_ast_visitor(gpr_to_sympy):
+    """parse GPR into SYMPY using ast and Node Visitor
+
+    Parameters
+    ----------
+    gpr_to_sympy : GPR
+        GPR which contains the gene reaction rule to parse
+
+    Returns
+    -------
+    Symbol or BooleanFunction
+        SYMPY expression (Symbol or And or Or). Symbol("") if the GPR is empty.
+    """
+    tree = gpr_to_sympy
+    if not tree.body:
+        return Symbol("")
+    sympifier = GPRSympifier()
+    sympy_exp = sympifier.visit(tree)
+    return sympy_exp
+
+
+def parse_sympy_via_names(gpr_to_sympy):
+    sympy_str = ast2sympy_str(gpr_to_sympy)
+    if sympy_str:
+        return eval(sympy_str)
```

Comment:
Didn't you say that this will fail for larger GPRs anyways?

Reply:
No. I'm pretty sure that this version will not crash, since the crashing version was using sympy.parse_expr

Reply:
Oh yeah sorry, I overlooked the previous part. But that seems a bit insecure due to the unprotected eval.